### ISSUE TITLE
net/http: don't reuse connections that are experiencing errors [1.20 backport]

### ISSUE THREAD

**Issue by btasker** [golang/go#60301]
Requesting that #59690 be considered for backport into the next 1.20 minor release

Relevant change in Gerrit is https://go-review.googlesource.com/c/net/+/486156

Rationale:

When encountered, this is a serious issue: requests are blocked until the kernel gives up on the socket (with default settings on a Linux host, that's around 17 minutes). If the operator restarts the application as a result, data loss may occur.

Although the fix is in `x/net` v0.10.0 it will not be available to most developers who likely import `net/http` (and so rely on the h2 bundle rolled into Go's releases)

The only viable workaround is to disable the H2 client entirely via environment variable.

**Comment by mdempsky** [comment]
@neild What do you think about this?

**Comment by mknyszek** [comment]
Friendly ping @neild. :)

**Comment by ianlancetaylor** [comment]
Opened #60662 for a 1.19 backport.

**Comment by Yuhjiang** [comment]
hello, [btasker](https://github.com/btasker), your description inspired me. I've also encountered the problem with go version `1.18.6`, and the problem would persist for 15-17min at every time. So I'm curious about the description of "with default settings on a Linux host, that's around 17 minutes". How can I get some related information about it?

**Comment by btasker** [comment]
Hi @Yuhjiang 

Sure. I don't know your level, so apologies if any of this is teaching you to suck eggs.

So, the bit you know:

1. Your application establishes a TCP connection to `$server` and negotiates to use `HTTP/2`
2. At some point, packets start getting lost and the connection breaks
3. The application keeps trying to send new requests over that connection, which fail
4. About 15 minutes later, a new connection is established and the application recovers

That 15 minute delay in recovery is a product of the way that the Linux kernel handles TCP retransmissions between 2 and 4 above.

When things are fully broken (i.e. the other end doesn't seem to be responding at all), there's an exponential back-off between each retry, for example
```
Send
wait 200ms
retry 1
wait 400ms
retry 2
wait 800ms
retry 3
wait 1.6s

etc
```

The minimum and maximum length of delay between the retries is defined by `TCP_RTO_MIN` (default 200ms) and `TCP_RTO_MAX` (default 120s).  These hardcoded into the kernel (although you can technically change them with BPF) [here](https://github.com/torvalds/linux/blob/692b7dc87ca6d55ab254f8259e6f970171dc9d01/include/net/tcp.h#L141).

The number of retries, however, is governed by a sysctl - `net.ipv4.tcp_retries2`. The default value is 15 (so with each pause doubling in length, you get up to that 15 minute mark).

Eventually, the limit in `net.ipv4.tcp_retries2` is hit and the kernel decides the connection is dead and closes it (causing it to drop out of the application's connection pool, with a new connection getting opened instead).

### Mitigation

So, if you're looking to mitigate this through system config changes, you could do so via  `net.ipv4.tcp_retries2`. If you drop it to 8 the maximum connection hang time will be around 1 minute (which still isn't great, but is in line with timeout thresholds that often get set at application level).

If you wanted to go a step further and adjust `TCP_RTO_MIN` and `TCP_RTO_MAX` using BPF there's (what looks like) a good write up at [https://arthurchiao.art/blog/customize-tcp-initial-rto-with-bpf/](https://arthurchiao.art/blog/customize-tcp-initial-rto-with-bpf/).


Does that help? Happy to go into more (or less) depth if needed

**Comment by Yuhjiang** [comment]
@btasker Thank you very much for you detailed reply, I learn a lot from it. I've figured out the problem and the reason. Finally, you've been a big help. I will try `x/net@v0.10.0` soon to verify if it could solve the problem.

**Comment by dmitshur** [comment]
Issue #60818 may also be relevant here.

**Comment by dmitshur** [comment]
Given the report in #60818 of this fix causing a problem, moving back to CherryPickCandidate state so we can take another look.

**Comment by alkmc** [comment]
@dmitshur seems that https://github.com/golang/go/issues/60818 is closed now.
Perhaps it is the right time to look again into this issue?

**Comment by dmitshur** [comment]
The upstream issue isn't resolved yet, and Go 1.20 has exited its support window per https://go.dev/doc/devel/release#policy.
